### ISSUE TITLE
Lint: `unused_assignments` lint not triggered when assigning though `Option` on a `Copy` type

### ISSUE THREAD

**Issue by epbuennig** [rust-lang/rust#102770]
When matching over a `&mut Option` and binding the result to a `mut PATTERN` instead of a `ref mut PATTTERN` or `PATTERN`, the resulting pattern is moved out, for copy types this will usually result in an `unused_assignments` lint.

Given the following code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1a5bc95859328b95d21023136c0c743f)), this doesn't happen, despite the local copy never being read after being assigned.
Changing the Type `A` to not being `Copy` correctly gives an error that the pattern attempting to move out.

```rust
#[derive(Debug, PartialEq)]
struct Container(Option<A>);

#[derive(Debug, Clone, Copy, PartialEq)]
struct A(Option<i32>);

impl Container {
    fn f(&mut self, int: i32) {
        if let Some(mut inner) = &mut self.0 {
            // inner: &mut A
            // mut inner: mut A
            // ref mut inner: &mut A
            // this should trigger a unused_assignments lint as inner is copied out and written but never read
            inner.0 = Some(int);
        }
    }
}

fn main() {
    let mut container = Container(Some(A(None)));
    container.f(42);

    // this fails as the actual variable was not set as expected
    assert_eq!(container, Container(Some(A(Some(42)))));
}
```

The current output is:

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
```

Ideally the output should look like:

```
warning: variable `inner` is assigned to, but never used
  --> src/main.rs:9:25
   |
22 |         if let Some(mut inner) = &mut self.0 {
   |                         ^^^^^
   |
   = note: `#[warn(unused_variables)]` on by default
   = note: consider using `_inner` instead

warning: value assigned to `inner.0` is never read
  --> src/main.rs:13:13
   |
23 |            inner.0 = Some(int);
   |            ^^^^^^^
   |
   = note: `#[warn(unused_assignments)]` on by default
   = help: maybe it is overwritten before being read?

warning: `playground` (bin "playground") generated 2 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
```
This was tested on:
```
rustc 1.66.0-nightly (f5193a9fc 2022-09-25)
binary: rustc
commit-hash: f5193a9fcc73dc09e41a90c5a2c97fc9acc16032
commit-date: 2022-09-25
host: x86_64-unknown-linux-gnu
release: 1.66.0-nightly
LLVM version: 15.0.0
```
and on:
```
rustc 1.66.0-nightly (f5193a9fc 2022-09-25)
binary: rustc
commit-hash: f5193a9fcc73dc09e41a90c5a2c97fc9acc16032
commit-date: 2022-09-25
host: x86_64-unknown-linux-gnu
release: 1.66.0-nightly
LLVM version: 15.0.0
```

**Comment by cjgillot** [comment]
This will be fixed by https://github.com/rust-lang/rust/pull/101500 if it lands.
